File: public/dsa/two_pointers/dutch_national_flag/optimal/dutch_national_flag.cpp
#include <bits/stdc++.h>
using namespace std;

void sortTheArray(vector<int> &A) {
    int low = 0, mid = 0, high = A.size() - 1;
    while (mid <= high) {
        if (A[mid] == 0) {
            swap(A[low], A[mid]);
            low++;
            mid++;
        } else if (A[mid] == 1) {
            mid++;
        } else {  // nums[mid] == 2
            swap(A[mid], A[high]);
            high--;
        }
    }
}

// tests-start
int main() {
    vector<int> A = {2, 2, 0, 1};
    sortTheArray(A);
    for (int v : A) {
        cout << v << " ";
    }
    cout << endl;
    return 0;
}
// tests-end

/*output
0 1 2 2
*/

File: public/dsa/two_pointers/dutch_national_flag/optimal/dutch_national_flag.py

from typing import List


class Solution:
    def sortTheArray(self, A: List[int]) -> None:
        low, mid, high = 0, 0, len(A) - 1

        while mid <= high:
            if A[mid] == 0:
                A[low], A[mid] = A[mid], A[low]
                low += 1
                mid += 1
            elif A[mid] == 1:
                mid += 1
            else:  # nums[mid] == 2
                A[mid], A[high] = A[high], A[mid]
                high -= 1


# tests-start
A = [2, 2, 0, 1]
sol = Solution()
sol.sortTheArray(A)
print(A)
# tests-end

'''output
[0, 1, 2, 2]
'''
